### ISSUE TITLE
100x Performance Regression from 3.1

### ISSUE THREAD

**Issue by pvande** [mruby/mruby#6369]
Given this program:
``` ruby
x = 400_000.times.map do |i|
  { f: 100.times.map { i.to_f }, a: [i] * 100, s: "a" * 100 }
end
```

Since 3.1.0, the runtime has increased over 100x.

``` sh
% time ./bin/mruby-3.1 main.rb
./bin/mruby-3.1 main.rb  10.04s user 0.31s system 97% cpu 10.622 total
% time ./bin/mruby-master main.rb
./bin/mruby-master main.rb  1013.02s user 1.32s system 97% cpu 17:15.59 total
```

(`mruby-master` was compiled from 12264e33e59e69daed0d313e418f38be7339f540 with `CFLAGS="-D MRB_ARY_LENGTH_MAX=0"`.)

I'll see if I can't bisect where the regression was introduced.

**Comment by pvande** [comment]
0c475708debc1465655dd58695cc776b223c9627 seems to be the commit that introduced this performance regression.

**Comment by matz** [comment]
It was caused by improper fix for #5791. I will address it soon.
